pkg/apis/core/types.go

```diff
@@ -1856,6 +1856,60 @@ type ClusterTrustBundleProjection struct {
 	Path string
 }
 
+// PodCertificateProjection provides a private key and X.509 certificate in
+// a combined file.
+type PodCertificateProjection struct {
+	// Kubelet's generated CSRs will be addressed to this signer.
+	SignerName string
+
+	// The type of keypair Kubelet will generate for the pod.
+	//
+	// Valid values are "RSA3072", "RSA4096", "ECDSAP256", "ECDSAP384", and
+	// "ED25519".  If left empty, Kubelet defaults to "ECDSAP256".
+	KeyType string
+
+	// What is the maximum lifetime of the cert that should be issued?  The
+	// actual issued certificate may have any lifetime *less* than
+	// maxExpirationSeconds.
+	//
+	// This value is copied into the PodCertificateRequest that kubelet creates
+	// to request the certificate.  See the documentation of
+	// PodCertificateRequest.spec.maxExpirationSeconds for information on how
+	// kube-apiserver and admission plugins process this value.
+	MaxExpirationSeconds *int32
+
+	// Write the credential bundle at this path in the projected volume.
```

Comment:
Can you please link original discussion.  When I see mutually exclusive sets of options, I tend to think discriminated union, not a series of fields where you must pick the right combination.  Best choice seems to be simply allowing these to be specified independently.  Doing so would also allow someone to specify all the options during a migration inside their binary.

Reply:
Done, they can now be specified independently.  We only check that at least one path is set.